Kernel/Devices/HID/VirtIO/Input.cpp

```diff
@@ -0,0 +1,469 @@
+    { Key_F12, 0xff },               { Key_Invalid, 0xff },           { Key_Invalid, 0xff },           { Key_Super, 0xff },
+    { Key_Invalid, 0xff },           { Key_Menu, 0xff },              { Key_Invalid, 0xff },           { Key_Invalid, 0xff },
+});
+// clang-format on
+
+UNMAP_AFTER_INIT NonnullLockRefPtr<Input> Input::must_create_for_pci_instance(PCI::DeviceIdentifier const& device_identifier)
+{
+    auto pci_transport_link = MUST(PCIeTransportLink::create(device_identifier));
+    return adopt_lock_ref_if_nonnull(new Input(move(pci_transport_link))).release_nonnull();
+}
+
+UNMAP_AFTER_INIT ErrorOr<void> Input::initialize_virtio_resources()
+{
+    TRY(Device::initialize_virtio_resources());
+    auto const* cfg = TRY(transport_entity().get_config(VirtIO::ConfigurationType::Device));
+    TRY(negotiate_features([&](auto) { return 0; }));
+
+    transport_entity().config_write8(*cfg, offsetof(VirtIOInputConfig, subsel), 0);
+
+    transport_entity().config_write8(*cfg, offsetof(VirtIOInputConfig, select), to_underlying(VirtIOInputConfig::Select::IDName));
+    transport_entity().read_config_atomic([&]() {
+        auto size = transport_entity().config_read8(*cfg, offsetof(VirtIOInputConfig, size));
+
+        VERIFY(size <= sizeof(VirtIOInputConfig::u.string));
+
+        char* name_chars = nullptr;
+        auto name = MUST(KString::try_create_uninitialized(size, name_chars));
+
+        for (size_t i = 0; i < size; i++)
+            name_chars[i] = static_cast<char>(transport_entity().config_read8(*cfg, offsetof(VirtIOInputConfig, u.string) + i));
+
+        dbgln("VirtIO::Input: Device name: {}", name);
+    });
+
+    VirtIOInputConfig::DevIDs dev_ids;
+
+    transport_entity().config_write8(*cfg, offsetof(VirtIOInputConfig, select), to_underlying(VirtIOInputConfig::Select::IDDevIDs));
+    transport_entity().read_config_atomic([&]() {
+        auto size = transport_entity().config_read8(*cfg, offsetof(VirtIOInputConfig, size));
+        VERIFY(size == sizeof(VirtIOInputConfig::u.ids));
+
+        dev_ids.bustype = bit_cast<LittleEndian<u16>>(transport_entity().config_read16(*cfg, offsetof(VirtIOInputConfig, u.ids.bustype)));
+        dev_ids.vendor = bit_cast<LittleEndian<u16>>(transport_entity().config_read16(*cfg, offsetof(VirtIOInputConfig, u.ids.vendor)));
+        dev_ids.product = bit_cast<LittleEndian<u16>>(transport_entity().config_read16(*cfg, offsetof(VirtIOInputConfig, u.ids.product)));
+        dev_ids.version = bit_cast<LittleEndian<u16>>(transport_entity().config_read16(*cfg, offsetof(VirtIOInputConfig, u.ids.version)));
+    });
+
+    dbgln("VirtIO::Input: DevIDs: bustype={:#x}, vendor={:#x}, product={:#x}, version={:#x}", dev_ids.bustype, dev_ids.vendor, dev_ids.product, dev_ids.version);
+
+    // Use the dev_ids to determine if the device is a QEMU mouse or keyboard, as we don't support devices sending both types of input events.
+    if (dev_ids.bustype == 0x6 && dev_ids.vendor == 0x627) {
```

Comment:
With #24210 merged, we could remove this devid matching code and instead always create a MouseDevice and KeyboardDevice for each virtio input device, which might be preferable?

Reply:
:shrug: Seems like a reasonable refactor after that architecture PR is ready to rock n roll.